src/DotNetWorker/Hosting/FunctionsApplicationBuilder.cs

```diff
@@ -0,0 +1,134 @@
+using System.Collections.Generic;
+using Microsoft.Azure.Functions.Worker;
+using Microsoft.Azure.Functions.Worker.Hosting;
+using Microsoft.Azure.Functions.Worker.Middleware;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Diagnostics.Metrics;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Extensions.Hosting;
+
+/// <summary>
+/// Placeholder
+/// </summary>
+public class FunctionsApplicationBuilder : IHostApplicationBuilder, IFunctionsWorkerApplicationBuilder
+{
+    private readonly HostApplicationBuilder _hostApplicationBuilder;
+    private readonly BootstrapHostBuilder _bootstrapHostBuilder;
+    private readonly IFunctionsWorkerApplicationBuilder _functionsWorkerApplicationBuilder;
+
+    private bool _hasBuilt = false;
+
+    internal const string SkipDefaultWorkerMiddlewareKey = "__SkipDefaultWorkerMiddleware";
+
+    internal FunctionsApplicationBuilder(string[]? args)
+    {
+        var configuration = new ConfigurationManager();
+
+        _hostApplicationBuilder = new HostApplicationBuilder(new HostApplicationBuilderSettings
+        {
+            Configuration = configuration
+        });
+
+        _bootstrapHostBuilder = new BootstrapHostBuilder(_hostApplicationBuilder);
+
+        // We'll set the default middleware up just before we build.
+        _bootstrapHostBuilder.Context.Properties[SkipDefaultWorkerMiddlewareKey] = true;
+
+        _bootstrapHostBuilder
+                   .ConfigureDefaults(args)
+                   .ConfigureFunctionsWorkerDefaults();
+
+        _functionsWorkerApplicationBuilder = InitializeHosting(_bootstrapHostBuilder);
+    }
+
+    /// <inheritdoc/>
+    public IDictionary<object, object> Properties => ((IHostApplicationBuilder)_hostApplicationBuilder).Properties;
+
+    IConfigurationManager IHostApplicationBuilder.Configuration => Configuration;
+
+    /// <inheritdoc/>
+    public ConfigurationManager Configuration => _hostApplicationBuilder.Configuration;
+
+    /// <inheritdoc/>
+    public IHostEnvironment Environment { get; private set; }
+
+    /// <inheritdoc/>
+    public ILoggingBuilder Logging => _hostApplicationBuilder.Logging;
+
+    /// <inheritdoc/>
+    public IMetricsBuilder Metrics => _hostApplicationBuilder.Metrics;
+
+    /// <inheritdoc />
+    public IServiceCollection Services => _hostApplicationBuilder.Services;
+
+    /// <summary>
+    /// Only available to AspNetCore extension (and maybe not even there in the future)
+    /// </summary>
+    internal IHostBuilder HostBuilder => _bootstrapHostBuilder;
+
+    /// <summary>
+    /// Placeholder
+    /// </summary>
+    /// <returns>placeholder</returns>
+    public IHost Build()
+    {
+        if (_hasBuilt)
+        {
+            throw new InvalidOperationException("The application has already been built.");
+        }
+
+        // we skipped this; make sure we add it at the end
+        this.UseDefaultWorkerMiddleware();
```

Comment:
What guarantee do we have that it was skipped?

Slightly related: in ASP.NET Core we use static properties on the `IApplicatonBuilder` to track what middleware has already been registered by user code to avoid re-registering. I assume something like this is not necessary of `UseDefaultWorkerMiddleware` is idempotent.

Reply:
> What guarantee do we have that it was skipped?
I guess we don't really have a guarantee, it's a good point. We need to figure out a good way to make sure that these are registered in the correct place, even if called at the wrong time. Even if we "locked" the middleware after this was called and threw if someone tried registering something else, it'd at least give us a guarantee that this made it to the right position. I'll come back to this in a follow-up (we've technically had this problem for years and I just ran into it with this PR).

> I assume something like this is not necessary of UseDefaultWorkerMiddleware is idempotent.
The problem here wasn't with re-registering as we do check if these have already bene registered. The problem was with the ordering of the registration.

Because now we're changing how we compose all the wire-ups, we call `ConfigureFunctionsWorkerDefaults()` twice now with the AspNetCore extension, which generally is fine. But this call to add the default middleware meant that our terminating (function execution) middleware was in the wrong spot -- not at the end. It'd get added at the first call, then all the AspNetCore middleware was added *after* it. This meant that we couldn't reach any of the middleware to do all the fun AspNetCore stuff.